File: sdk/csharp/Examples/SubscriptionModule.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;
using InstantDB;

/// <summary>
/// Subscription Module Example
///
/// This example demonstrates:
/// - Implementing subscription filter functions
/// - Transforming subscription events
/// - Providing initial data for subscriptions
/// - Client-specific event handling
/// - Using both context-based (reducers) and global (subscriptions) database access
/// </summary>
namespace SubscriptionModule
{
    // =============================================================================
    // Data Structures
    // =============================================================================

    /// <summary>
    /// User activity table schema
    /// </summary>
    [Table(Name = "user_activities")]
    public class UserActivity
    {
        [PrimaryKey]
        [JsonPropertyName("id")]
        public string Id { get; set; } = "";

        [JsonPropertyName("user_id")]
        public string UserId { get; set; } = "";

        [JsonPropertyName("action")]
        public string Action { get; set; } = "";

        [JsonPropertyName("timestamp")]
        public ulong Timestamp { get; set; }

        [JsonPropertyName("details")]
        public string Details { get; set; } = "";

        [JsonPropertyName("metadata")]
        public Dictionary<string, object> Metadata { get; set; } = new();
    }

    /// <summary>
    /// Activity event for real-time subscriptions
    /// </summary>
    [Event(Name = "activity_events")]
    public class ActivityEvent
    {
        [JsonPropertyName("activity_id")]
        public string ActivityId { get; set; } = "";

        [JsonPropertyName("user_id")]
        public string UserId { get; set; } = "";

        [JsonPropertyName("action")]
        public string Action { get; set; } = "";

        [JsonPropertyName("timestamp")]
        public ulong Timestamp { get; set; }

        [JsonPropertyName("event_type")]
        public string EventType { get; set; } = "";
    }

    /// <summary>
    /// User profile schema
    /// </summary>
    [Table(Name = "users")]
    public class User
    {
        [PrimaryKey]
        [JsonPropertyName("id")]
        public string Id { get; set; } = "";

        [JsonPropertyName("name")]
        public string Name { get; set; } = "";

        [JsonPropertyName("premium")]
        public bool Premium { get; set; }

        [JsonPropertyName("reputation")]
        public int Reputation { get; set; }

        [JsonPropertyName("created_at")]
        public ulong CreatedAt { get; set; }
    }

    // =============================================================================
    // Subscription Module Implementation
    // =============================================================================

    /// <summary>
    /// Main subscription module class with filters and transforms
    /// </summary>
    public class SubscriptionModuleImpl : ModuleBase
    {
        // =============================================================================
        // Subscription Functions
        // =============================================================================

        /// <summary>
        /// Filter function for recent activities (last hour only)
        /// </summary>
        [SubscriptionFilter(Name = "filter_recent_activities")]
        public static bool FilterRecentActivities(byte[] eventData)
        {
            try
            {
                var eventJson = System.Text.Encoding.UTF8.GetString(eventData);
                var eventObj = JsonSerializer.Deserialize<JsonElement>(eventJson);

                // Skip DELETE events
                if (!eventObj.TryGetProperty("new_value", out var newValueElement))
                {
                    return false;
                }

                var activityJson = newValueElement.GetString();
                if (string.IsNullOrEmpty(activityJson))
                {
                    return false;
                }

                var activity = JsonSerializer.Deserialize<UserActivity>(activityJson);
                var currentTime = Utils.Now();
                var oneHourMs = 60 * 60 * 1000UL;

                // Only include activities from the last hour
                return (currentTime - activity.Timestamp) <= oneHourMs;
            }
            catch (Exception ex)
            {
                Utils.LogError($"Filter error: {ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// Filter function for premium users only
        /// </summary>
        [SubscriptionFilter(Name = "filter_premium_users")]
        public static bool FilterPremiumUsers(byte[] eventData)
        {
            try
            {
                var eventJson = System.Text.Encoding.UTF8.GetString(eventData);
                var eventObj = JsonSerializer.Deserialize<JsonElement>(eventJson);

                if (!eventObj.TryGetProperty("new_value", out var newValueElement))
                {
                    return false;
                }

                var activityJson = newValueElement.GetString();
                if (string.IsNullOrEmpty(activityJson))
                {
                    return false;
                }

                var activity = JsonSerializer.Deserialize<UserActivity>(activityJson);

                // Check if user has premium account
                var userResult = DB.Read<User>(new Key(activity.UserId));
                if (userResult.IsErr || userResult.Unwrap() == null)
                {
                    return false;
                }

                var user = userResult.Unwrap();
                return user.Premium;
            }
            catch (Exception ex)
            {
                Utils.LogError($"Premium filter error: {ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// Transform function to add metadata and computed fields
        /// </summary>
        [SubscriptionTransform(Name = "transform_add_metadata")]
        public static byte[] TransformAddMetadata(byte[] eventData)
        {
            try
            {
                var eventJson = System.Text.Encoding.UTF8.GetString(eventData);
                var eventObj = JsonSerializer.Deserialize<JsonElement>(eventJson);

                if (eventObj.TryGetProperty("new_value", out var newValueElement))
                {
                    var activityJson = newValueElement.GetString();
                    if (!string.IsNullOrEmpty(activityJson))
                    {
                        var activity = JsonSerializer.Deserialize<UserActivity>(activityJson);

                        // Add computed metadata
                        activity.Metadata["processed_at"] = Utils.Now();
                        activity.Metadata["server_version"] = "1.0.0";

                        // Add user reputation score
                        var userResult = DB.Read<User>(new Key(activity.UserId));
                        if (userResult.IsOk && userResult.Unwrap() != null)
                        {
                            var user = userResult.Unwrap();
                            activity.Metadata["user_reputation"] = user.Reputation;
                            activity.Metadata["user_premium"] = user.Premium;
                        }

                        // Update the event with transformed data
                        var transformedJson = JsonSerializer.Serialize(activity);
                        var newEventObj = JsonSerializer.Deserialize<Dictionary<string, object>>(eventJson);
                        newEventObj["new_value"] = transformedJson;

                        var resultJson = JsonSerializer.Serialize(newEventObj);
                        return System.Text.Encoding.UTF8.GetBytes(resultJson);
                    }
                }

                return eventData; // Return original if no transformation needed
            }
            catch (Exception ex)
            {
                Utils.LogError($"Transform error: {ex.Message}");
                return eventData; // Return original on error
            }
        }

        /// <summary>
        /// Transform function to anonymize sensitive data
        /// </summary>
        [SubscriptionTransform(Name = "transform_anonymize")]
        public static byte[] TransformAnonymize(byte[] eventData)
        {
            try
            {
                var eventJson = System.Text.Encoding.UTF8.GetString(eventData);
                var eventObj = JsonSerializer.Deserialize<JsonElement>(eventJson);

                if (eventObj.TryGetProperty("new_value", out var newValueElement))
                {
                    var activityJson = newValueElement.GetString();
                    if (!string.IsNullOrEmpty(activityJson))
                    {
                        var activity = JsonSerializer.Deserialize<UserActivity>(activityJson);

                        // Replace user_id with anonymous hash
                        var hash = activity.UserId.GetHashCode();
                        activity.UserId = $"anon_{Math.Abs(hash % 100000)}";

                        // Remove potentially sensitive details
                        activity.Details = "[REDACTED]";

                        // Clear sensitive metadata
                        activity.Metadata.Clear();
                        activity.Metadata["anonymized"] = true;

                        var transformedJson = JsonSerializer.Serialize(activity);
                        var newEventObj = JsonSerializer.Deserialize<Dictionary<string, object>>(eventJson);
                        newEventObj["new_value"] = transformedJson;

                        var resultJson = JsonSerializer.Serialize(newEventObj);
                        return System.Text.Encoding.UTF8.GetBytes(resultJson);
                    }
                }

                return eventData;
            }
            catch (Exception ex)
            {
                Utils.LogError($"Anonymize error: {ex.Message}");
                return eventData;
            }
        }

        /// <summary>
        /// Get initial data for activity subscriptions
        /// </summary>
        [InitialData(Name = "get_initial_activities")]
        public static byte[] GetInitialActivities(string whereClause)
        {
            try
            {
                var initialData = new List<UserActivity>();

                // For demo purposes, create some sample activities
                // In a real implementation, you'd query the database based on the whereClause

                var activity1 = new UserActivity
                {
                    Id = "activity_1",
                    UserId = "user_123",
                    Action = "login",
                    Timestamp = Utils.Now() - 30000, // 30 seconds ago
                    Details = "User logged in from mobile app"
                };

                var activity2 = new UserActivity
                {
                    Id = "activity_2",
                    UserId = "user_456",
                    Action = "purchase",
                    Timestamp = Utils.Now() - 60000, // 1 minute ago
                    Details = "Purchased premium subscription"
                };

                initialData.Add(activity1);
                initialData.Add(activity2);

                var resultJson = JsonSerializer.Serialize(initialData);
                return System.Text.Encoding.UTF8.GetBytes(resultJson);
            }
            catch (Exception ex)
            {
                Utils.LogError($"Initial data error: {ex.Message}");
                return Array.Empty<byte>();
            }
        }

        // =============================================================================
        // Reducer Functions
        // =============================================================================

        /// <summary>
        /// Create a new user activity record
        /// </summary>
        [Reducer]
        public static int CreateActivity(ReducerContext ctx, string userId, string action, string details)
        {
            try
            {
                // Validate inputs
                if (string.IsNullOrWhiteSpace(userId))
                {
                    Utils.LogError("User ID cannot be empty");
                    return -1;
                }

                if (string.IsNullOrWhiteSpace(action))
                {
                    Utils.LogError("Action cannot be empty");
                    return -1;
                }

                var activity = new UserActivity
                {
                    Id = $"{userId}_{Utils.Now()}",
                    UserId = userId,
                    Action = action,
                    Timestamp = Utils.Now(),
                    Details = details ?? ""
                };

                // Insert into database (automatically triggers changefeed)
                ctx.Db.GetTable<UserActivity>().Insert(activity);

                // Note: Changefeed events are now automatically emitted by the database layer

                Utils.LogInfo($"Created activity: {activity.Id} ({action})");
                return 1;
            }
            catch (Exception ex)
            {
                Utils.LogError($"Exception in CreateActivity: {ex.Message}");
                return -1;
            }
        }

        /// <summary>
        /// Get user activities with filtering
        /// </summary>
        [Reducer]
        public static int GetUserActivities(ReducerContext ctx, string userId, long sinceTimestamp)
        {
            try
            {
                Utils.LogInfo($"Getting activities for user: {userId} since: {sinceTimestamp}");

                // In a real implementation, you'd scan the activities table
                // and filter by user_id and timestamp
                // For now, just return success
                return 1;
            }
            catch (Exception ex)
            {
                Utils.LogError($"Exception in GetUserActivities: {ex.Message}");
                return -1;
            }
        }

        /// <summary>
        /// Create a test user for demo purposes
        /// </summary>
        [Reducer]
        public static int CreateTestUser(ReducerContext ctx, string userId, string name, bool premium, int reputation)
        {
            try
            {
                var user = new User
                {
                    Id = userId,
                    Name = name,
                    Premium = premium,
                    Reputation = reputation,
                    CreatedAt = Utils.Now()
                };

                // Insert into database (automatically triggers changefeed)
                ctx.Db.GetTable<User>().Insert(user);

                Utils.LogInfo($"Created test user: {userId} ({name})");
                return 1;
            }
            catch (Exception ex)
            {
                Utils.LogError($"Exception in CreateTestUser: {ex.Message}");
                return -1;
            }
        }

        // =============================================================================
        // Module Lifecycle Overrides
        // =============================================================================

        public override int Initialize()
        {
            Utils.LogInfo("Subscription module initialized successfully");

            // Note: In the new API, initialization would use a system context
            // For now, initialization is simpler without complex database operations

            return 0;
        }

        public override int OnClientConnected(string connectionId)
        {
            Utils.LogInfo($"Client connected to subscription module: {connectionId}");

            // Note: Client lifecycle events could trigger reducer calls with system context
            // For now, keep these simple without database operations

            return 0;
        }

        public override int OnClientDisconnected(string connectionId)
        {
            Utils.LogInfo($"Client disconnected from subscription module: {connectionId}");

            // Note: Client lifecycle events could trigger reducer calls with system context
            // For now, keep these simple without database operations

            return 0;
        }
    }
}

File: sdk/cpp/examples/subscription_module.cpp
#include "../instantdb.hpp"
#include <string>
#include <nlohmann/json.hpp>

using namespace instantdb;

/**
 * Subscription Module Example
 *
 * This example demonstrates:
 * - Implementing subscription filter functions
 * - Transforming subscription events
 * - Providing initial data for subscriptions
 * - Client-specific event emission
 */

// =============================================================================
// Data Structures
// =============================================================================

struct UserActivity {
    std::string user_id;
    std::string action;
    uint64_t timestamp;
    std::string details;

    static std::vector<uint8_t> serialize(const UserActivity& activity) {
        nlohmann::json j;
        j["user_id"] = activity.user_id;
        j["action"] = activity.action;
        j["timestamp"] = activity.timestamp;
        j["details"] = activity.details;

        std::string json_str = j.dump();
        return std::vector<uint8_t>(json_str.begin(), json_str.end());
    }

    static Result<UserActivity> deserialize(const std::vector<uint8_t>& data) {
        try {
            std::string json_str(data.begin(), data.end());
            nlohmann::json j = nlohmann::json::parse(json_str);

            UserActivity activity;
            activity.user_id = j["user_id"];
            activity.action = j["action"];
            activity.timestamp = j["timestamp"];
            activity.details = j["details"];

            return activity;
        } catch (const std::exception& e) {
            return Error("Failed to deserialize UserActivity: " + std::string(e.what()));
        }
    }
};

// Specialize serialization for UserActivity
namespace instantdb::serialization {

template<>
class Serializable<UserActivity> {
public:
    static std::vector<uint8_t> serialize(const UserActivity& value) {
        return UserActivity::serialize(value);
    }

    static Result<UserActivity> deserialize(const std::vector<uint8_t>& data) {
        return UserActivity::deserialize(data);
    }
};

} // namespace instantdb::serialization

// =============================================================================
// Subscription Functions
// =============================================================================

/**
 * Filter function for user activity subscriptions
 * Only include activities from the last hour
 */
INSTANTDB_EXPORT bool filter_activity_by_time(const std::vector<uint8_t>& event_data) {
    try {
        // Parse the changefeed event
        std::string json_str(event_data.begin(), event_data.end());
        nlohmann::json event = nlohmann::json::parse(json_str);

        // Extract timestamp from the event
        if (!event.contains("new_value")) {
            return false; // Skip DELETE events
        }

        // Parse the activity data
        std::string activity_json = event["new_value"];
        nlohmann::json activity = nlohmann::json::parse(activity_json);

        uint64_t activity_time = activity["timestamp"];
        uint64_t current_time = utils::now();
        uint64_t one_hour_ms = 60 * 60 * 1000;

        // Only include activities from the last hour
        return (current_time - activity_time) <= one_hour_ms;

    } catch (const std::exception& e) {
        utils::log(utils::LogLevel::Error, "Filter error: " + std::string(e.what()));
        return false;
    }
}

/**
 * Filter function for specific user activities
 * Only include activities for users with premium accounts
 */
INSTANTDB_EXPORT bool filter_premium_users(const std::vector<uint8_t>& event_data) {
    try {
        std::string json_str(event_data.begin(), event_data.end());
        nlohmann::json event = nlohmann::json::parse(json_str);

        if (!event.contains("new_value")) {
            return false;
        }

        std::string activity_json = event["new_value"];
        nlohmann::json activity = nlohmann::json::parse(activity_json);

        std::string user_id = activity["user_id"];

        // Check if user has premium account
        auto user_result = db::read<nlohmann::json>("users", user_id);
        if (user_result.is_err() || !user_result.unwrap().has_value()) {
            return false;
        }

        auto user = user_result.unwrap().value();
        return user.contains("premium") && user["premium"].get<bool>();

    } catch (const std::exception& e) {
        utils::log(utils::LogLevel::Error, "Premium filter error: " + std::string(e.what()));
        return false;
    }
}

/**
 * Transform function to add computed fields
 */
INSTANTDB_EXPORT std::vector<uint8_t> transform_add_metadata(const std::vector<uint8_t>& event_data) {
    try {
        std::string json_str(event_data.begin(), event_data.end());
        nlohmann::json event = nlohmann::json::parse(json_str);

        if (event.contains("new_value")) {
            std::string activity_json = event["new_value"];
            nlohmann::json activity = nlohmann::json::parse(activity_json);

            // Add computed metadata
            activity["processed_at"] = utils::now();
            activity["server_version"] = "1.0.0";

            // Add user reputation score
            std::string user_id = activity["user_id"];
            auto user_result = db::read<nlohmann::json>("users", user_id);
            if (user_result.is_ok() && user_result.unwrap().has_value()) {
                auto user = user_result.unwrap().value();
                if (user.contains("reputation")) {
                    activity["user_reputation"] = user["reputation"];
                }
            }

            // Update the event with transformed data
            event["new_value"] = activity.dump();
        }

        std::string result_json = event.dump();
        return std::vector<uint8_t>(result_json.begin(), result_json.end());

    } catch (const std::exception& e) {
        utils::log(utils::LogLevel::Error, "Transform error: " + std::string(e.what()));
        return event_data; // Return original on error
    }
}

/**
 * Transform function to anonymize sensitive data
 */
INSTANTDB_EXPORT std::vector<uint8_t> transform_anonymize(const std::vector<uint8_t>& event_data) {
    try {
        std::string json_str(event_data.begin(), event_data.end());
        nlohmann::json event = nlohmann::json::parse(json_str);

        if (event.contains("new_value")) {
            std::string activity_json = event["new_value"];
            nlohmann::json activity = nlohmann::json::parse(activity_json);

            // Replace user_id with anonymous hash
            if (activity.contains("user_id")) {
                std::string user_id = activity["user_id"];
                activity["user_id"] = "anon_" + std::to_string(std::hash<std::string>{}(user_id) % 100000);
            }

            // Remove potentially sensitive details
            if (activity.contains("details")) {
                activity["details"] = "[REDACTED]";
            }

            event["new_value"] = activity.dump();
        }

        std::string result_json = event.dump();
        return std::vector<uint8_t>(result_json.begin(), result_json.end());

    } catch (const std::exception& e) {
        utils::log(utils::LogLevel::Error, "Anonymize error: " + std::string(e.what()));
        return event_data;
    }
}

/**
 * Get initial data for activity subscriptions
 */
INSTANTDB_EXPORT std::vector<uint8_t> get_initial_activities(const std::string& where_clause) {
    try {
        nlohmann::json initial_data = nlohmann::json::array();

        // For demo purposes, create some sample activities
        // In a real implementation, you'd query the database based on the where_clause

        nlohmann::json activity1;
        activity1["user_id"] = "user_123";
        activity1["action"] = "login";
        activity1["timestamp"] = utils::now() - 30000; // 30 seconds ago
        activity1["details"] = "User logged in from mobile app";

        nlohmann::json activity2;
        activity2["user_id"] = "user_456";
        activity2["action"] = "purchase";
        activity2["timestamp"] = utils::now() - 60000; // 1 minute ago
        activity2["details"] = "Purchased premium subscription";

        initial_data.push_back(activity1);
        initial_data.push_back(activity2);

        std::string result_json = initial_data.dump();
        return std::vector<uint8_t>(result_json.begin(), result_json.end());

    } catch (const std::exception& e) {
        utils::log(utils::LogLevel::Error, "Initial data error: " + std::string(e.what()));
        return {};
    }
}

// =============================================================================
// Reducer Functions
// =============================================================================

/**
 * Create a new user activity record
 */
INSTANTDB_EXPORT int32_t create_activity(const char* user_id, const char* action, const char* details) {
    try {
        UserActivity activity{
            .user_id = std::string(user_id),
            .action = std::string(action),
            .timestamp = utils::now(),
            .details = std::string(details)
        };

        std::string activity_id = std::string(user_id) + "_" + std::to_string(activity.timestamp);

        auto result = db::write("user_activities", activity_id, activity);
        if (result.is_err()) {
            utils::log(utils::LogLevel::Error, "Failed to create activity: " + result.error().message());
            return -1;
        }

        // Emit real-time event
        auto emit_result = events::emit("activity_created", activity_id, activity);
        if (emit_result.is_err()) {
            utils::log(utils::LogLevel::Warn, "Failed to emit activity event: " + emit_result.error().message());
        }

        utils::log(utils::LogLevel::Info, "Created activity: " + activity_id + " (" + action + ")");
        return 1;

    } catch (const std::exception& e) {
        utils::log(utils::LogLevel::Error, "Exception in create_activity: " + std::string(e.what()));
        return -1;
    }
}

/**
 * Get user activities with filtering
 */
INSTANTDB_EXPORT int32_t get_user_activities(const char* user_id, int64_t since_timestamp) {
    try {
        utils::log(utils::LogLevel::Info,
                  "Getting activities for user: " + std::string(user_id) +
                  " since: " + std::to_string(since_timestamp));

        // In a real implementation, you'd scan the activities table
        // and filter by user_id and timestamp
        // For now, just return success
        return 1;

    } catch (const std::exception& e) {
        utils::log(utils::LogLevel::Error, "Exception in get_user_activities: " + std::string(e.what()));
        return -1;
    }
}

// =============================================================================
// Module Lifecycle
// =============================================================================

INSTANTDB_INIT() {
    utils::log(utils::LogLevel::Info, "Subscription module initialized");

    // Create some sample users for testing
    nlohmann::json user1;
    user1["id"] = "user_123";
    user1["name"] = "Alice";
    user1["premium"] = true;
    user1["reputation"] = 95;

    nlohmann::json user2;
    user2["id"] = "user_456";
    user2["name"] = "Bob";
    user2["premium"] = false;
    user2["reputation"] = 72;

    db::write("users", "user_123", user1);
    db::write("users", "user_456", user2);

    return 0;
}

INSTANTDB_CLIENT_CONNECTED() {
    utils::log(utils::LogLevel::Info, "Client connected to subscription module");
    return 0;
}

INSTANTDB_CLIENT_DISCONNECTED() {
    utils::log(utils::LogLevel::Info, "Client disconnected from subscription module");
    return 0;
}